### PR TITLE
Progress Bar Implementation + Refactoring View Architecture + UI Changes

### PR DESCRIPTION
**Pull request by tkv29**
This pull request (https://github.com/bptlab/TracEX/issues/67) introduces following features:

1. Progress Bar
Added a progress bar for the extraction pipeline which uses a JS script to pull the current progress of session variables. 
For the journey generation a spinning loading wheel was added due to the relativley short waiting time and very small amount of steps.

2. View Architecture Changes
- an separate filter view was added, to remove duplicate code
- patient journey generator was changed to a script instead of using an orchestrator with one module
- patient journey genrator overview and result were merged into one view, while the logic behind the patient journey generator was outsourced into another view. This allows more flexibility in journey generation and the ability to direct from the filter view to the patient journey result view
image

3. UI Changes
Added a page transition effect for being fancy

### CODE REVIEW

File: tracex/tracex/views.py [hunk 1]
```diff
@@ -1,6 +1,14 @@
...
+class TracexLandingPage(TemplateView):
+    """View for the landing page of the tracex app."""
+
+    template_name = "landing_page.html"
+
+    def get_context_data(self, **kwargs):
+        """Reset session variables."""
+        context = super().get_context_data(**kwargs)
+        self.request.session.flush()
+        return context
```
Comment (FR-SON) [inline tracex/tracex/views.py:14]:
```suggestion
        self.request.session.flush()
        
        return context
```

File: tracex/extraction/forms.py [hunk 2]
```diff
@@ -87,26 +87,8 @@ class Meta:
...
-        }
-        widgets = {
-            "name": forms.TextInput(
-                attrs={"placeholder": "Name for your patient journey"}
-            ),
-        }
-
-    field_order = ["name", "event_types", "locations"]
-
+class FilterForm(BaseEventForm):
```
Comment (FR-SON) [inline tracex/extraction/forms.py:90]:
```suggestion

class FilterForm(BaseEventForm):
```

File: tracex/extraction/forms.py [hunk 3]
```diff
@@ -87,26 +87,8 @@ class Meta:
...
+    """Form for selecting filter for extraction result"""
 
 class ResultForm(BaseEventForm):
```
Comment (FR-SON) [inline tracex/extraction/forms.py:93]:
```suggestion

class ResultForm(BaseEventForm):
```

**Review by FR-SON — APPROVED** [review]
The progress bar is not rendered correctly…when using Firefox or Safari. Chrome works though. Looks amazing.

**Review by FR-SON — CHANGES_REQUESTED** [review]
i suggest some minor changes

File: tracex/extraction/views.py [hunk 4]
```diff
@@ -27,18 +27,24 @@ class JourneyInputView(generic.CreateView):
...
     form_class = JourneyForm
     template_name = "upload_journey.html"
-    success_url = reverse_lazy("processing")
+    success_url = reverse_lazy("journey_filter")
+
+    def get_context_data(self, **kwargs):
+        """Clear session variables"""
+        self.request.session["is_extracted"] = False
+        self.request.session["progress"] = 0
+        self.request.session["status"] = None
```
[the remark this answers is not in the document]
Comment (tkv29) [reply tracex/extraction/views.py:36]: it is used so we can display on the front end the current executed module

File: tracex/extraction/views.py [hunk 5]
```diff
@@ -27,18 +27,24 @@ class JourneyInputView(generic.CreateView):
...
+        self.request.session.save()
```
[the remark this answers is not in the document]
Comment (tkv29) [reply tracex/extraction/views.py:37]: It is necessary so we can ensure that the are no old values. In some cases the js query is faster then the orchestrator can update the session variables. Therefore we need to clean them beforehand. But actually it should be called when entering the filter view. I changed that

**Comment by tkv29** [comment]
> i suggest some minor changes

I replied to some comments and implement your change requests

**Comment by tkv29** [comment]
> i suggest some minor changes
@FR-SON 
I replied to some comments and implement your change requests

File: tracex/extraction/logic/orchestrator.py [hunk 6]
```diff
@@ -102,21 +102,31 @@ def initialize_modules(self):
         print("Initialization of modules successful.")
         return modules
 
-    def run(self):
+    def run(self, view=None):
         """Run the modules."""
         modules = self.initialize_modules()
+        modules_number = len(modules) + 2
```
Comment (PitButtchereit) [inline tracex/extraction/logic/orchestrator.py:108]: There should be no hardcoding here. Why is this necessary right now?

File: tracex/extraction/logic/orchestrator.py [hunk 7]
```diff
@@ -102,21 +102,31 @@ def initialize_modules(self):
...
+            )
+            current_step += 1
 
+        self.update_progress(view, current_step, modules_number, "Cohort Tagger")
         self.db_objects["cohort"] = self.configuration.modules[
             "cohort_tagging"
         ]().execute_and_save(self.data, patient_journey)
+        current_step += 1
         for module in modules:
+            self.update_progress(view, current_step, modules_number, module.name)
```
Comment (PitButtchereit) [inline tracex/extraction/logic/orchestrator.py:126]: The maximum number of modules is required once but passed for each execution of a module. Although we a only talking about a number here, it is generally discouraged to pass arguments multiple times if they don't change and are only required one time. Try to fix this.

File: tracex/extraction/static/extraction/js/progress-bar.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+                setTimeout(updateProgressBar, 1000); // Fetch progress again after 1 second
+            }
+        },
+        error: function() {
+            console.log("Error occurred while fetching progress.");
+        }
+    });
+}
+
+function getProgressBarColor(progress) {
```
Comment (PitButtchereit) [inline tracex/extraction/static/extraction/js/progress-bar.js:38]: Maybe just change the progress bar to one color. There color has no indication for us and colors like "bg-danger" do not sound very suitable in our case.

File: tracex/extraction/views.py [hunk 9]
```diff
@@ -47,14 +44,52 @@ def form_valid(self, form):
...
 
-    def get(self, request, *args, **kwargs):
-        """Redirect to result page."""
-        return redirect("result")
+    form_class = FilterForm
+    template_name = "filter_journey.html"
+    success_url = reverse_lazy("result")
+
+    def form_valid(self, form):
+        """Save the uploaded journey in the cache."""
```
Comment (PitButtchereit) [inline tracex/extraction/views.py:56]: I'm not realy confident with django views but the docstring from this method suggests to me that when i validate the form the journey will be saved to the cache and something else happens like that i am redirected to the next view. However in the code below you can see that the run method from the orchestrator is called hence our pipeline is started. Can you make the code more clear here?

**Review by PitButtchereit — CHANGES_REQUESTED** [review]

**Comment by tkv29** [comment]
Update to the browser support problem:
I identified that the problem lies within the functionality of the submit button "Execute Extraction Pipeline". With the click on the button the page will reload and interrupt the ajax script. Browsers based on Chromium can handle that and still executing the script. We should tackle this problem in the future due to the lower priority. In meanwhile I added a spinning wheel if the progress bar is not working.

File: tracex/extraction/logic/orchestrator.py (hunk 6 above)
Comment (PitButtchereit) [reply tracex/extraction/logic/orchestrator.py]: Sure, I know what the number means. It is just that when were are adding new modules in the future we'll have to think about this hard coded number or we'll be wondering and debugging why the progress bar is not displaying correctly

Comment (tkv29) [reply tracex/extraction/logic/orchestrator.py]: @FR-SON Is it possible that the other two modules are being executed part of the pipeline?

File: tracex/extraction/logic/orchestrator.py (hunk 7 above)
Comment (tkv29) [reply tracex/extraction/logic/orchestrator.py]: The only way is to use a class variable but I'm don't think that this would be the right use case.

Comment (tkv29) [reply tracex/extraction/logic/orchestrator.py]: I change module_number to be an instance variable. But in retrospective its doesnt change anything at all. The passing arguments doesnt reduce and instead of passing a number we are passing a whole instance.

File: tracex/extraction/logic/orchestrator.py (hunk 6 above)
Comment (FR-SON) [reply tracex/extraction/logic/orchestrator.py]: We could use `self.configuration.modules`. Every Module that is used in the orchestrator should be listed there.

**Comment by FR-SON** [comment]
grafik

The `get` function in the `FIlterView` and `JourneyGenerationView` seem to have some kind of issue. 
This is what the base method looks like:
grafik

File: tracex/extraction/logic/orchestrator.py (hunk 7 above)
Comment (FR-SON) [reply tracex/extraction/logic/orchestrator.py]: If we use the modules list from the configuration instead (like i suggested in the other thread we don`t need that variable at all. we can access `len(self.configuration.modules)` from within the `update_progress()` function.

If we use the instance variable, it must be defined in the `__init__` function. Something like `self.modules_number: int = 0`.

File: tracex/extraction/logic/orchestrator.py (hunk 6 above)
Comment (tkv29) [reply tracex/extraction/logic/orchestrator.py]: sounds good, thx!

File: tracex/extraction/logic/orchestrator.py (hunk 7 above)
Comment (tkv29) [reply tracex/extraction/logic/orchestrator.py]: @FR-SON
I'm not sure if this solves the "problem". To access len(self.configuration.modules) we have to pass "self" to the function. So in the end we passing the whole instance everytime instead of one number, I guess

Comment (tkv29) [reply tracex/extraction/logic/orchestrator.py]: I think the only solution is to create a class variable if we really want to reduce the amount of passing arguments (but not sure if thats the right use case)

File: tracex/extraction/logic/orchestrator.py [hunk 10]
```diff
@@ -151,3 +161,11 @@ def save_results_to_db(self):
         trace.save()
         patient_journey.trace.add(trace)
         patient_journey.save()
+
+    def update_progress(self, view, current_step, modules_number, module_name):
```
[the remark this answers is not in the document]
Comment (tkv29) [reply tracex/extraction/logic/orchestrator.py]: notice: this suggestions is not realized based on the new implementation using update_progress as a method of the class

File: tracex/extraction/logic/orchestrator.py (hunk 7 above)
Comment (tkv29) [reply tracex/extraction/logic/orchestrator.py]: @FR-SON thx for the suggestion, in the end I think it doesnt solves the problem to reduce the amount of passing parameters. Instead of passing a number we are passing the whole instance, so its basically just an indirection. 

But this solution is definitely better and already implemented.

**Comment by tkv29** [comment]
@FR-SON fix is provided

**Review by FR-SON — APPROVED** [review]
